File: tools/s3_sink_reference/pkg/s3/writer.go
package s3

import (
	"fmt"
	"math/rand"
	"time"
)

type Part struct {
	PartNumber int
	ETag       string
}

// Writer is a simplified S3 client interface
type Writer interface {
	InitMultipart(key string) (string, error)
	UploadPart(uploadID string, partNumber int, data []byte) (string, error)
	CompleteMultipart(uploadID string, parts []Part) error
	PutObject(key string, data []byte) error
}

// MockWriter simulates S3 behavior with latency
type MockWriter struct {
	Bucket string
	Region string
}

func NewMockWriter(bucket, region string) *MockWriter {
	return &MockWriter{Bucket: bucket, Region: region}
}

func (w *MockWriter) InitMultipart(key string) (string, error) {
	// Simulate latency
	time.Sleep(50 * time.Millisecond)
	uploadID := fmt.Sprintf("upload_id_%d", rand.Int63())
	fmt.Printf("[S3] InitMultipart: %s/%s -> %s\n", w.Bucket, key, uploadID)
	return uploadID, nil
}

func (w *MockWriter) UploadPart(uploadID string, partNumber int, data []byte) (string, error) {
	// Simulate upload time (10MB/s)
	sizeMB := float64(len(data)) / (1024 * 1024)
	latency := time.Duration(sizeMB * 100) * time.Millisecond
	time.Sleep(latency)
	
	etag := fmt.Sprintf("etag_%d", partNumber)
	fmt.Printf("[S3] UploadPart: %s (Part %d, %d bytes) -> %s\n", uploadID, partNumber, len(data), etag)
	return etag, nil
}

func (w *MockWriter) CompleteMultipart(uploadID string, parts []Part) error {
	time.Sleep(100 * time.Millisecond)
	fmt.Printf("[S3] CompleteMultipart: %s (%d parts)\n", uploadID, len(parts))
	return nil
}

func (w *MockWriter) PutObject(key string, data []byte) error {
	time.Sleep(100 * time.Millisecond)
	fmt.Printf("[S3] PutObject: %s/%s (%d bytes)\n", w.Bucket, key, len(data))
	return nil
}

File: src/io/sink/writer.zig
const std = @import("std");
const xev = @import("xev");
const s3_proto = @import("../../../protocol/s3.zig");

// TODO: Import transport/connection infrastructure
// For now, we define the Interface expected by the Sink Task.
// Real implementation will wire up to src/io/s3/global_pool.zig or similar.

pub const S3Writer = struct {
    allocator: std.mem.Allocator,
    bucket: []const u8,
    key: []const u8,
    region: []const u8,
    // pool: *GlobalPool, // TODO

    pub fn init(allocator: std.mem.Allocator, bucket: []const u8, key: []const u8, region: []const u8) S3Writer {
        return .{
            .allocator = allocator,
            .bucket = bucket,
            .key = key,
            .region = region,
        };
    }

    /// Initiate a multipart upload. Returns the Upload ID.
    /// Caller owns the returned string.
    pub fn initMultipartUpload(self: *S3Writer) ![]const u8 {
        // Mock implementation for scaffold
        // Real impl:
        // 1. Format request using s3_proto.formatInitiateMultipartRequest
        // 2. Send via transport
        // 3. Parse XML using s3_proto.parseUploadId

        // Return dummy ID
        return self.allocator.dupe(u8, "dummy_upload_id_123");
    }

    /// Upload a part. Returns the ETag.
    /// Caller owns the returned string.
    pub fn uploadPart(self: *S3Writer, upload_id: []const u8, part_number: u32, data: []const u8) ![]const u8 {
        // Mock implementation
        // Real impl:
        // 1. Format using s3_proto.formatUploadPartRequest
        // 2. Send payload
        // 3. Extract ETag from response headers

        // Simulate IO
        std.time.sleep(10 * std.time.ns_per_ms);
        return self.allocator.dupe(u8, "dummy_etag");
    }

    /// Complete multipart upload.
    pub fn completeMultipartUpload(self: *S3Writer, upload_id: []const u8, parts: []const s3_proto.S3.Part) !void {
        // Mock implementation
        // Real impl:
        // 1. Format using s3_proto.formatCompleteMultipartRequest
        // 2. Send XML body
    }

    /// Single PUT object (for small files).
    pub fn putObject(self: *S3Writer, data: []const u8) !void {
        // Mock implementation
    }
};
